zephyr/lib/alloc.c

```diff
@@ -189,6 +196,86 @@ static void l3_heap_free(struct k_heap *h, void *mem)
 
 #endif
 
+#if CONFIG_VIRTUAL_HEAP
+static void *virtual_heap_alloc(struct vmh_heap *heap, uint32_t flags, uint32_t caps, size_t bytes,
+				uint32_t align)
+{
+	void *mem;
+
+	K_SPINLOCK(&vmh_lock) {
+		heap->core_id = cpu_get_id();
+		mem = vmh_alloc(heap, bytes);
+	}
+
+	if (!mem)
+		return NULL;
+
+	assert(IS_ALIGNED(mem, align));
+
+	if (flags & SOF_MEM_FLAG_COHERENT)
+		return sys_cache_uncached_ptr_get((__sparse_force void __sparse_cache *)mem);
+
+	return mem;
+}
+
+/**
+ * Checks whether pointer is from virtual memory range.
+ * @param ptr Pointer to memory being checked.
+ * @return True if pointer falls into virtual memory region, false otherwise.
+ */
+static bool is_virtual_heap_pointer(void *ptr)
+{
+	uintptr_t virtual_heap_start = POINTER_TO_UINT(sys_cache_cached_ptr_get(&heapmem)) +
+				       HEAPMEM_SIZE;
+	uintptr_t virtual_heap_end = CONFIG_KERNEL_VM_BASE + CONFIG_KERNEL_VM_SIZE;
+
+	if (!is_cached(ptr))
+		ptr = sys_cache_cached_ptr_get((__sparse_force void *)ptr);
+
+	return ((POINTER_TO_UINT(ptr) >= virtual_heap_start) &&
+		(POINTER_TO_UINT(ptr) < virtual_heap_end));
+}
+
+static void virtual_heap_free(void *ptr)
+{
+	ptr = sys_cache_cached_ptr_get((__sparse_force void *)ptr);
+
+	K_SPINLOCK(&vmh_lock) {
+		vmh_free(virtual_buffers_heap, ptr);
+	}
+}
+
+static const struct vmh_heap_config static_hp_buffers = {
+	{
+		{ 128, 32},
+		{ 512, 8},
+		{ 1024, 16},
+		{ 2048, 8},
+		{ 4096, 11},
+		{ 8192, 13},
+		{ 65536, 3},
+		{ 131072, 1},
+		{ 524288, 1} /* buffer for kpb */
```

Comment:
Long term do we want these mappings to come from topology or extended manifest ? We had this sort of hard mappings with the IPC3 allocator and there was always a frequent need to update and target mappings based on the use case/topology.

Reply:
@lgirdwood Ultimately, we plan this division will be configured in runtime via ipc4. However, it is not critical because it splits the virtual address space (16 MB). Physical memory blocks are only mapped during allocation. In the current version, the allocator has quite restrictive requirements for space division. The buffer size must be a power of two and each buffer bundle must be a multiple of the memory page size.